src/libsystemd-network/sd-dhcp-lease.c

```diff
@@ -559,6 +573,123 @@ static int lease_parse_sip_server(const uint8_t *option, size_t len, struct in_a
         return lease_parse_in_addrs(option + 1, len - 1, ret, n_ret);
 }
 
+static int lease_parse_dns_name(const uint8_t *optval, size_t optlen, char **ret) {
+        _cleanup_free_ char *name = NULL;
+        int r;
+
+        assert(optval);
+        assert(ret);
+
+        r = dns_name_from_wire_format(&optval, &optlen, &name);
+        if (r < 0)
+                return r;
+        if (r == 0 || optlen != 0)
+                return -EBADMSG;
+
+        *ret = TAKE_PTR(name);
+        return r;
+}
+
+static int lease_parse_dnr(const uint8_t *option, size_t len, sd_dns_resolver **dnr, size_t *n_dnr) {
+        int r;
+        sd_dns_resolver *res_list = NULL;
+        size_t n_resolvers = 0;
+        CLEANUP_ARRAY(res_list, n_resolvers, sd_dns_resolver_array_free);
+
+        assert(option);
+        assert(dnr);
+
+        _cleanup_(sd_dns_resolver_done) sd_dns_resolver res = {};
+
+        size_t offset = 0;
+        while (offset < len) {
+                if (len < 2)
+                        return -EBADMSG;
+
+                /* Instance Data length */
+                size_t ilen = unaligned_read_be16(&option[offset]);
```

Comment:
Looks like it goes sideways here:
```sh
=================================================================
==1076333==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x615000000741 at pc 0x00000057762a bp 0x7ffd63f90df0 sp 0x7ffd63f90de8
READ of size 2 at 0x615000000741 thread T0
    #0 0x577629 in unaligned_read_be16 /home/vagrant/systemd/build/../src/basic/unaligned.h:14:27
    #1 0x577629 in lease_parse_dnr /home/vagrant/systemd/build/../src/libsystemd-network/sd-dhcp-lease.c:610:31
    #2 0x56f8c0 in dhcp_lease_parse_options /home/vagrant/systemd/build/../src/libsystemd-network/sd-dhcp-lease.c:1010:21
    #3 0x55c677 in parse_options /home/vagrant/systemd/build/../src/libsystemd-network/dhcp-option.c:345:33
    #4 0x55bda8 in dhcp_option_parse /home/vagrant/systemd/build/../src/libsystemd-network/dhcp-option.c:373:13
    #5 0x55605f in client_parse_message /home/vagrant/systemd/build/../src/libsystemd-network/sd-dhcp-client.c:1477:13
    #6 0x542778 in client_handle_offer_or_rapid_ack /home/vagrant/systemd/build/../src/libsystemd-network/sd-dhcp-client.c:1562:13
    #7 0x541eb5 in LLVMFuzzerTestOneInput /home/vagrant/systemd/build/../src/libsystemd-network/fuzz-dhcp-client.c:85:13
    #8 0x442b64 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/vagrant/systemd/build/fuzz-dhcp-client+0x442b64) (BuildId: 92086e5395a952f37c8b59d3fbbc523a7fbc9247)
    #9 0x42ab37 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/home/vagrant/systemd/build/fuzz-dhcp-client+0x42ab37) (BuildId: 92086e5395a952f37c8b59d3fbbc523a7fbc9247)
    #10 0x4305f6 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/vagrant/systemd/build/fuzz-dhcp-client+0x4305f6) (BuildId: 92086e5395a952f37c8b59d3fbbc523a7fbc9247)
    #11 0x45cee6 in main (/home/vagrant/systemd/build/fuzz-dhcp-client+0x45cee6) (BuildId: 92086e5395a952f37c8b59d3fbbc523a7fbc9247)
    #12 0x7f7bc3649b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #13 0x7f7bc3649c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #14 0x425144 in _start (/home/vagrant/systemd/build/fuzz-dhcp-client+0x425144) (BuildId: 92086e5395a952f37c8b59d3fbbc523a7fbc9247)

0x615000000741 is located 0 bytes after 449-byte region [0x615000000580,0x615000000741)
allocated by thread T0 here:
    #0 0x4f7a72 in malloc (/home/vagrant/systemd/build/fuzz-dhcp-client+0x4f7a72) (BuildId: 92086e5395a952f37c8b59d3fbbc523a7fbc9247)
    #1 0x7f7bc58b56cb in operator new(unsigned long) (/lib64/libstdc++.so.6+0xb56cb) (BuildId: 694f64a512ecd55e04f1587ddd790ec30cb0b726)
    #2 0x42ab37 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/home/vagrant/systemd/build/fuzz-dhcp-client+0x42ab37) (BuildId: 92086e5395a952f37c8b59d3fbbc523a7fbc9247)
    #3 0x4305f6 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/vagrant/systemd/build/fuzz-dhcp-client+0x4305f6) (BuildId: 92086e5395a952f37c8b59d3fbbc523a7fbc9247)
    #4 0x45cee6 in main (/home/vagrant/systemd/build/fuzz-dhcp-client+0x45cee6) (BuildId: 92086e5395a952f37c8b59d3fbbc523a7fbc9247)
    #5 0x7f7bc3649b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #6 0x7f7bc3649c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #7 0x425144 in _start (/home/vagrant/systemd/build/fuzz-dhcp-client+0x425144) (BuildId: 92086e5395a952f37c8b59d3fbbc523a7fbc9247)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/vagrant/systemd/build/../src/basic/unaligned.h:14:27 in unaligned_read_be16
Shadow bytes around the buggy address:
  0x615000000480: 00 00 00 00 00 00 00 00 01 fa fa fa fa fa fa fa
  0x615000000500: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x615000000580: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x615000000600: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x615000000680: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x615000000700: 00 00 00 00 00 00 00 00[01]fa fa fa fa fa fa fa
  0x615000000780: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x615000000800: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x615000000880: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x615000000900: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x615000000980: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 02
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1076333==ABORTING
```
The option was
```
          "dhcp.option.type": "162",
          "dhcp.option.type_tree": {
            "dhcp.option.length": "55",
            "dhcp.option.value": "00:19:00:01:09:03:61:62:63:03:78:79:7a:00:04:01:02:03:04:00:01:00:04:03:64:6f:74:00:19:00:01:09:03:78:79:7a:03:61:62:63:00:04:05:06:07:08:00:01:00:04:03:64:6f:74:00"
          }
```
I think it could have been triggered by CIFuzz but as far as I can see the `dnr_v4` file comes with a few DHCP options without the BOOTP part and that prevents the fuzz target from catching this fast enough. I prepended the BOOTP part to it
```
00000000  01 01 06 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000000e0  00 00 00 00 00 00 00 00  00 00 00 00 63 82 53 63  |............c.Sc|
000000f0  02 00 06 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000100  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000001e0  00 00 00 00 a2 36 00 19  00 01 09 03 61 62 63 03  |.....6......abc.|
000001f0  78 79 7a 00 04 01 02 03  04 00 01 00 04 03 64 6f  |xyz...........do|
00000200  74 00 19 00 01 09 03 78  79 7a 03 61 62 63 00 04  |t......xyz.abc..|
00000210  05 06 07 08 00 01 00 04  03 64 6f 74              |.........dot|
0000021c
```

Reply:
Can you check if the fixup resolves this? Think I missed this spot when reworking this function last.

> I think it could have been triggered by CIFuzz but as far as I can see the dnr_v4 file comes with a few DHCP options without the BOOTP part and that prevents the fuzz target from catching this fast enough.

Oh, good to know. Do you know if the dnr_v6 one formatted correctly?

Reply:
Looks like with the last commit applied the fuzz target no longer fails with that option. I think once `dnr_v4` is updated the CI should be able to confirm that too.

Reply:
Applied, thanks.